### PR TITLE
Fix mounting of vfat on *BSD

### PR DESCRIPTION
**Pull request by igalic**
## Proposed Commit Message
Fix mounting of vfat filesystems by normalizing the different names for vfat to "msdos" which works across BSDS.

## Additional Context

FreeBSD Bugzilla ref: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=250496
LP# https://bugs.launchpad.net/cloud-init/+bug/1901958

## Test Steps

This doesn't reproduce on Ubuntu.

## Checklist:
 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

### CODE REVIEW

File: cloudinit/util.py [hunk 1]
```diff
@@ -1667,10 +1667,12 @@ def mount_cb(device, callback, data=None, mtype=None,
             mtypes = ["auto"]
     elif platsys.endswith("bsd"):
         if mtypes is None:
-            mtypes = ['ufs', 'cd9660', 'vfat']
+            mtypes = ['ufs', 'cd9660', 'msdos']
         for index, mtype in enumerate(mtypes):
             if mtype == "iso9660":
                 mtypes[index] = "cd9660"
+            if mtype in ["vfat", "msdosfs", "msdos"]:
```
Comment (OddBloke) [inline cloudinit/util.py:1674]: We don't need `"msdos"` in this list; `mtypes[index]` is already `"msdos"` in that case.

```suggestion
            if mtype in ["vfat", "msdosfs"]:
```

File: cloudinit/tests/test_util.py [hunk 2]
```diff
@@ -730,6 +730,36 @@ def already_mounted_device(self, already_mounted_device_and_mountdict):
...
+    @mock.patch("cloudinit.util.subp.subp")
+    @mock.patch("cloudinit.temp_utils.tempdir")
+    def test_normalize_vfat_on_bsd(self, m_tmpdir, m_subp, m_is_BSD):
+        m_is_BSD.return_value = "SomeBSD"
+        m_tmpdir.__enter__ = mock.Mock(return_value="/tmp/fake")
+        m_tmpdir.__exit__ = mock.Mock(return_value=True)
+        callback = mock.Mock()
+
+        util.mount_cb('/dev/fake0', callback, mtype='vfat')
+        m_subp.assert_called_with(
```
Comment (OddBloke) [inline cloudinit/tests/test_util.py]: If we're using `assert_*` methods then we need `autospec=True` on the corresponding patch call; (search for "autospec" in https://cloudinit.readthedocs.io/en/latest/topics/hacking.html#testing).

File: cloudinit/tests/test_util.py [hunk 3]
```diff
@@ -730,6 +730,36 @@ def already_mounted_device(self, already_mounted_device_and_mountdict):
...
+    @mock.patch("cloudinit.util.subp.subp")
+    @mock.patch("cloudinit.temp_utils.tempdir")
+    def test_normalize_iso9660_on_bsd(self, m_tmpdir, m_subp, m_is_BSD):
+        m_is_BSD.return_value = "SomeBSD"
+        m_tmpdir.__enter__ = mock.Mock(return_value="/tmp/fake")
+        m_tmpdir.__exit__ = mock.Mock(return_value=True)
+        callback = mock.Mock()
+
+        util.mount_cb('/dev/fake0', callback, mtype='iso9660')
+        m_subp.assert_called_with(
```
Comment (OddBloke) [inline cloudinit/tests/test_util.py]: Same here.

File: cloudinit/tests/test_util.py [hunk 4]
```diff
@@ -730,6 +730,36 @@ def already_mounted_device(self, already_mounted_device_and_mountdict):
         """already_mounted_device_and_mountdict, but return only the device"""
         return already_mounted_device_and_mountdict[0]
 
+    @mock.patch("platform.system")
+    @mock.patch("cloudinit.util.subp.subp")
+    @mock.patch("cloudinit.temp_utils.tempdir")
+    def test_normalize_vfat_on_bsd(self, m_tmpdir, m_subp, m_is_BSD):
+        m_is_BSD.return_value = "SomeBSD"
```
Comment (OddBloke) [inline cloudinit/tests/test_util.py:737]: Should this just be `= True`?

File: cloudinit/tests/test_util.py [hunk 5]
```diff
@@ -730,6 +730,36 @@ def already_mounted_device(self, already_mounted_device_and_mountdict):
         """already_mounted_device_and_mountdict, but return only the device"""
         return already_mounted_device_and_mountdict[0]
 
+    @mock.patch("platform.system")
+    @mock.patch("cloudinit.util.subp.subp")
+    @mock.patch("cloudinit.temp_utils.tempdir")
+    def test_normalize_vfat_on_bsd(self, m_tmpdir, m_subp, m_is_BSD):
```
Comment (OddBloke) [inline cloudinit/tests/test_util.py:736]: This test could be expanded via parameterisation to also exercise the "msdosfs" codepath (and perhaps also combined with the below test). Let me know if you'd like more details and I can run you through it!

**Review by OddBloke — CHANGES_REQUESTED** [review]
Thanks for this, @igalic! The way these tests are structured and the boundaries you have chosen to test along LGTM (as does the actual change itself!). I have some inline comments to clean up some details.

File: cloudinit/tests/test_util.py [hunk 6]
```diff
@@ -730,6 +730,36 @@ def already_mounted_device(self, already_mounted_device_and_mountdict):
         """already_mounted_device_and_mountdict, but return only the device"""
         return already_mounted_device_and_mountdict[0]
 
+    @mock.patch("platform.system")
```
[the remark this answers is not in the document]
Comment (igalic) [reply cloudinit/tests/test_util.py:733]: yeah, i thought i could get away with cheating ;)

**Review by OddBloke — APPROVED** [review]
LGTM, thanks!

**Pull request merged**